File: naiveBayesClassifier/src/main/java/org/metplus/curriculum/cruncher/naivebayes/CruncherImpl.java
package org.metplus.curriculum.cruncher.naivebayes;

import de.daslaboratorium.machinelearning.classifier.Classification;
import de.daslaboratorium.machinelearning.classifier.bayes.BayesClassifier;
import org.metplus.curriculum.cruncher.Cruncher;
import org.metplus.curriculum.cruncher.CruncherMetaData;
import org.metplus.curriculum.database.domain.MetaDataField;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

/**
 * Created by joao on 5/28/16.
 * Class that implements the Naive Bayes cruncher
 */
public class CruncherImpl implements Cruncher {
    private static final Logger logger = LoggerFactory.getLogger(CruncherImpl.class);
    public static final String CRUNCHER_NAME = "NaiveBayes";

    private BayesClassifier<String, String> classifier;

    public void setCleanExpressions(List<String> cleanExpressions) {
        this.cleanExpressions = cleanExpressions;
    }
    public BayesClassifier<String, String> getClassifier() {
        return classifier;
    }

    private List<String> cleanExpressions;

    public CruncherImpl() {
        classifier = new BayesClassifier();
        this.cleanExpressions = new ArrayList<>();
    }
    public CruncherImpl(List<String> cleanExpressions) {
        classifier = new BayesClassifier();
        this.cleanExpressions = cleanExpressions;
    }

    /**
     * Crunch a string
     * @param data Input to be processed
     * @return Cruncher meta data
     */
    public CruncherMetaData crunch(String data) {
        logger.trace("crunch(" + data + ")");
        Collection<Classification<String, String>> results = classifier.classifyDetailed(tokenize(data));
        NaiveBayesMetaData metaData = generateMetaData(results);
        return metaData;
    }

    /**
     * Generate the meta data needed for a collection of classifications
     * @param classificationResult Result of a classification
     * @return Meta data ready to be saved
     */
    private NaiveBayesMetaData generateMetaData(Collection<Classification<String, String>> classificationResult) {
        logger.trace("generateMetaData(" + classificationResult + ")");
        NaiveBayesMetaData allMetaData = new NaiveBayesMetaData();
        float maxProbability = 0.0f;
        allMetaData.setBestMatchCategory(null);
        String output = "Cruncher categories: [";
        for(Classification<String, String> metaData: classificationResult) {
            if(allMetaData.getBestMatchCategory() == null)
                allMetaData.setBestMatchCategory(metaData.getCategory());
            else if(maxProbability < metaData.getProbability()) {
                allMetaData.setBestMatchCategory(metaData.getCategory());
                maxProbability = metaData.getProbability();
            }
            output += ("" + metaData.getCategory() + ": " + metaData.getProbability() + ", ");

            if(metaData.getProbability() <= 0)
                break;
            MetaDataField<Float> field = new MetaDataField<>(metaData.getProbability());
            allMetaData.addField(metaData.getCategory(), field);
            output += ("" + metaData.getCategory() + ": " + metaData.getProbability() + ", ");
        }
        logger.debug(output + "]");
        return allMetaData;
    }

    public String getCruncherName() {
        return CRUNCHER_NAME;
    }

    /**
     * Reset the memory of the classifier
     */
    public void resetMemory() {
        classifier.reset();
    }

    /**
     * Train the cuncher
     * @param database Database with all the features and their tokens
     */
    public void train(Map<String, List<String>> database) {
        logger.trace("train(" + database + ")");
        if(database != null && database.size() > 0 ) {
            logger.debug("Training using a database with {} categories", database.size());
            for (Map.Entry<String, List<String>> entry : database.entrySet()) {
                train(entry.getKey(), entry.getValue());
            }
        }
        logger.info("Train complete");
    }

    /**
     * Retrieve all the categories of the classifier
     * @return Set with category names
     */
    public Set<String> getCategories() {
        return classifier.getCategories();
    }

    /**
     * Train the cruncher
     * @param feature Feature name
     * @param trainningSet Set of tokens that represent a feature
     */
    public void train(String feature, List<String> trainningSet) {
        if(trainningSet == null)
            return;
        for(String text: trainningSet)
            classifier.learn(feature, tokenize(feature, text));
    }

    /**
     * Function used to tokenize a string
     * @param text To be tokenized
     * @return List of tokens from the text
     */
    private List<String> tokenize(String text) {
        return tokenize("", text);
    }

    /**
     * Tokenize using a feature
     * @param feature find a specific feature and save it as a token
     * @param text Text to be tokenized
     * @return List of tokens from the text
     */
    private List<String> tokenize(String feature, String text) {
        List<String> result = new ArrayList<>();
        String newText = text.replaceAll(feature, feature.replaceAll("\\s+", "@@@@")).toLowerCase();
        if(cleanExpressions.size() > 0) {
            for(String exp: cleanExpressions) {
                newText = newText.replaceAll("\\b" + exp + "\\b", " ");
            }
        }
        for(String expression: newText.split("\\s")) {
            if(expression.length() == 0)
                continue;
            String token = expression.replaceAll("@@@@", " ").replaceAll("[^a-z0-9 ]", "");
            result.add(token);
        }
        return result;
    }
}

File: crunchers/naiveBayes/src/main/kotlin/org/metplus/curriculum/cruncher/naivebayes/CruncherImpl.kt
package org.metplus.curriculum.cruncher.naivebayes

import de.daslaboratorium.machinelearning.classifier.Classification
import de.daslaboratorium.machinelearning.classifier.bayes.BayesClassifier
import org.metplus.cruncher.rating.Cruncher
import org.metplus.cruncher.rating.CruncherMetaData
import org.slf4j.LoggerFactory
import java.util.ArrayList

class CruncherImpl(
        private var cleanExpressions: List<String> = mutableListOf()
) : Cruncher {
    private val logger = LoggerFactory.getLogger(CruncherImpl::class.java)
    private val name = "NaiveBayes"
    val classifier = BayesClassifier<String, String>()

    override fun crunch(data: String): CruncherMetaData {
        logger.trace("crunch($data)")
        val results = classifier.classifyDetailed(tokenize(data))
        return generateMetaData(results).metaData
    }

    private fun generateMetaData(classificationResult: Collection<Classification<String, String>>): NaiveBayesMetaData {
        logger.trace("generateMetaData($classificationResult)")
        val allMetaData = NaiveBayesMetaData()
        var maxProbability = 0.0f
        allMetaData.bestMatchCategory = null
        var output = "Cruncher categories: ["
        for (metaData in classificationResult) {
            if (allMetaData.bestMatchCategory == null)
                allMetaData.bestMatchCategory = metaData.category
            else if (maxProbability < metaData.probability) {
                allMetaData.bestMatchCategory = metaData.category
                maxProbability = metaData.probability
            }

            if (metaData.probability <= 0)
                continue

            allMetaData.metaData.metaData[metaData.category] = metaData.probability.toDouble()
            output += "" + metaData.category + ": " + metaData.probability + ", "
        }
        allMetaData.totalProbability = maxProbability.toDouble()
        logger.debug(output + "], Best Match: " + allMetaData.bestMatchCategory + ": " + allMetaData.totalProbability)
        return allMetaData
    }

    override fun getCruncherName(): String {
        return name
    }

    fun train(database: Map<String, List<String>>) {
        logger.trace("train($database)")
        if (database.isNotEmpty()) {
            logger.debug("Training using a database with {} categories", database.size)
            for (entry in database.entries) {
                train(entry.key, entry.value)
            }
        }
        logger.info("Train complete")
    }

    fun train(feature: String, trainningSet: List<String>?) {
        if (trainningSet == null)
            return
        for (text in trainningSet)
            classifier.learn(feature, tokenize(feature, text))
    }

    private fun tokenize(text: String): List<String> {
        return tokenize("", text)
    }

    private fun tokenize(feature: String, text: String): List<String> {
        val result = ArrayList<String>()
        var newText = text.replace(feature.toRegex(), feature.replace("\\s+".toRegex(), "@@@@")).toLowerCase()
        if (cleanExpressions.isNotEmpty()) {
            for (exp in cleanExpressions) {
                newText = newText.replace("\\b$exp\\b".toRegex(), " ")
            }
        }
        for (expression in newText.split("\\s".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
            if (expression.trim { it <= ' ' }.isEmpty())
                continue
            val token = expression.replace("@@@@".toRegex(), " ").replace("[^a-z0-9 ]".toRegex(), "")
            if (token.trim { it <= ' ' }.isEmpty())
                continue
            result.add(token)
        }
        return result
    }
}